exquisite_corpus/spm_naive_language_model.py

```diff
@@ -0,0 +1,428 @@
+        a test dataset.  Assumes that the model outputs are log prababilities.
+        """
+        if device is not None:
+            self.device = device
+
+        self.model.eval()
+        test_data = SpmIdsDataset(test_dataset_path, min_length=min_length)
+        sampler = SpmIdsBatchSampler(test_data, batch_size=batch_size, randomize=False)
+        pin_memory = self.device != torch.device("cpu")
+        data_loader = DataLoader(
+            test_data,
+            batch_sampler=sampler,
+            pin_memory=pin_memory,
+            num_workers=n_data_loader_workers,
+            collate_fn=self.collate_batch_with_labels,
+        )
+        n_labels = torch.tensor(0, dtype=torch.int64, device=self.device)
+        minus_log_prob_sum = torch.tensor(0.0, dtype=torch.float32, device=self.device)
+        nll_loss = LossFunction(reduction="sum")
+        with torch.autograd.no_grad():
+            for batch, labels in data_loader:
+                batch = batch.to(self.device, non_blocking=True)
+                labels = labels.to(self.device, non_blocking=True)
+                log_probs = self.model(batch)
+                minus_log_prob_sum += nll_loss(log_probs, labels)
+                n_labels += labels.numel()
+                perplexity = (minus_log_prob_sum / n_labels).exp().item()
+                print(
+                    "Processed {} labels; running perplexity is {}.".format(
+                        n_labels, perplexity
+                    )
+                )
+        perplexity = (minus_log_prob_sum / n_labels).exp().item()
+        print("Perplexity on {} is {}.".format(test_dataset_path, perplexity))
+        return perplexity
+
+    def save(self, path):
+        # We deliberately don't save the device.
+        mgr_kwargs = dict(
+            n_tokens=self.model.encoder.num_embeddings,
+            n_dims=self.model.encoder.embedding_dim,
+            sparse=self.model.encoder.sparse,
+            hidden_size=self.model.rnn.hidden_size,
+            num_layers=self.model.rnn.num_layers,
+            bias=self.model.rnn.bias,
+            dropout=self.model.rnn.dropout,
+            bidirectional=self.model.rnn.bidirectional,
+        )
+        model_state_dict = self.model.state_dict()
+        torch.save([mgr_kwargs, model_state_dict], str(path))
+
+    @classmethod
+    def load(cls, path, device=torch.device("cuda:0")):
+        # In the past we've seen issues with excessive GPU memory consumption
+        # when loading models direct to GPU, so we load to CPU first then move.
+        mgr_kwargs, model_state_dict = torch.load(str(path), map_location="cpu")
+        mgr_kwargs.update(device=torch.device("cpu"))  # load to cpu
+        result = cls(**mgr_kwargs)
+        result.model.load_state_dict(model_state_dict)
+        result.device = device  # move to requested or saved device if any
+        return result
+
+
+def main(model_path=None, model_kwargs=None, task="train", task_kwargs=None):
```

Comment:
Two different dictionaries that become kwargs? What do they need to contain?

I'm hoping this isn't just what happens in idiomatic PyTorch code. It looks to me like this function is trying to be way too configurable.

I don't want a function that does everything when you pass it a carefully constructed pile of configuration. I want a function that does a specific thing that works.

Reply:
I've removed all config files and made the command line tool train a model (with default parameters) if invoked with no options.